tokio/src/sync/once_cell.rs

```diff
@@ -304,30 +352,144 @@ impl<T> OnceCell<T> {
     /// [`SetError::AlreadyInitializedError`]: crate::sync::SetError::AlreadyInitializedError
     /// [`SetError::InitializingError`]: crate::sync::SetError::InitializingError
     pub fn set(&self, value: T) -> Result<(), SetError<T>> {
-        if self.initialized() {
-            return Err(SetError::AlreadyInitializedError(value));
+        // special handling for ZSTs
+        if std::mem::size_of::<T>() == 0 {
+            return self.set_zst(value);
+        }
+        // Try to lock the state. Relaxed ordering is enough for success,
+        // as the value is written but not read after. In case of failure,
+        // acquire ordering is used to have happens-before the successful
+        // initialization.
+        match self.state.compare_exchange(
+            STATE_UNSET,
+            STATE_LOCKED,
+            Ordering::Relaxed,
+            Ordering::Acquire,
```

Comment:
I would expect the code below to be okay:
```rs
static mut GLOBAL: i32 = 0;
static GLOBAL_INIT: OnceCell<()> = OnceCell::new();
```
Imagine you have 10 threads do this in parallel:
```rs
let is_leader = cell.get_or_try_init(|| unsafe {
    GLOBAL += 1;
    if GLOBAL == 10 {
    	Ok(())
    } else {
    	Err(())
    }
}).is_ok();

if is_leader {
    assert_eq!(unsafe { GLOBAL }, 10);
}
```
But we need `Acquire` on `UNSET`->`LOCKED` and `Release` on `LOCKED`->`SET` for the above to be synchronized correctly. With the current relaxed orderings, two different initializers would have a data race on `GLOBAL`.

Reply:
I would not have thought to this use case ... To make it work, the memory order in the `DropGuard` should also be changed to `Release`. And in the same spirit, success memory order to the first `set` CAS should also be `Acquire`, to have an happens-before relation with potential previously failed `get_or_try_init`.

Reply:
Fixed in https://github.com/tokio-rs/tokio/pull/7161/commits/7a79f29d90876f87d059f6176f737c2876c41657

Reply:
I've added a loom test for this use case in https://github.com/tokio-rs/tokio/pull/7161/commits/33754c2bf3e2d9f58da8c15f96693d4b53364d96